### PR TITLE
[algo] fix: skip tool observations in RF++ reward-to-go

### PR DESCRIPTION
**Pull request by haoyang9804**
## What This Fixes

This fixes a quiet multi-turn training-signal bug in veRL's
`reinforce_plus_plus` advantage path.

In multi-turn/tool rollout, `response_ids` can contain both model output tokens
and tool observation tokens. The tool observation tokens are not trainable
actions, so their `response_mask` entries are `0`, but they are still internal
tokens in the same response trajectory.

For this constructed trajectory:

```text
token_level_rewards = [[0.0, 0.0, 1.0]]
response_mask       = [[1.0, 0.0, 1.0]]
gamma               = 0.5
tokens              = [model token, tool observation token, later model token]
```

the earlier model token should still receive discounted reward-to-go from the
later model token:

```text
fixed returns       = [[0.5, 1.0, 1.0]]
```

Before this patch, veRL treated the internal `response_mask=0` tool observation
as a terminal boundary and reset the running return:

```text
unpatched returns   = [[0.0, 0.5, 1.0]]
```

Training does not crash. The failure is that the model action before a tool
observation silently loses the downstream reward signal.

## Root Cause

`compute_reinforce_plus_plus_outcome_advantage()` scanned rewards from right to
left and reset `running_return` whenever `response_mask[:, t] == 0`.

That behavior is correct for padding after EOS in ordinary single-turn response
tensors. It is incorrect for multi-turn/tool rollout, because mask gaps can also
represent tool observations inside the response. Those observation tokens should
not contribute to the loss, but they should not break reward-to-go between model
actions.

## Fix

The reverse scan now only updates the running return on trainable model tokens.
For `response_mask=0` observation tokens, it leaves `running_return` unchanged
and writes the passthrough value into `returns`. The masked position remains
excluded from whitening and policy loss by the existing `response_mask`.

The regression test covers the minimal `[model, tool observation, model]`
boundary shape.

## Duplicate Check

I checked upstream before preparing this PR and did not find an existing matching
fix or report:

- PR search: [`reinforce_plus_plus multi-turn response_mask returns tool observation`](https://github.com/verl-project/verl/pulls?q=is%3Apr+reinforce_plus_plus+multi-turn+response_mask+returns+tool+observation)
- PR search: [`REINFORCE++ response_mask reward-to-go multi-turn`](https://github.com/verl-project/verl/pulls?q=is%3Apr+REINFORCE%2B%2B+response_mask+reward-to-go+multi-turn)
- PR search: [`reinforce_plus_plus running_return response_mask EOS`](https://github.com/verl-project/verl/pulls?q=is%3Apr+reinforce_plus_plus+running_return+response_mask+EOS)
- Issue search: [`reinforce_plus_plus response_mask reward-to-go tool observation`](https://github.com/verl-project/verl/issues?q=is%3Aissue+reinforce_plus_plus+response_mask+reward-to-go+tool+observation)

No matching upstream PR or issue was found as of 2026-05-17.

## Validation Recipe

```json
{
  "kind": "rl_sentinel_validation_recipe",
  "schema_version": 1,
  "bug_id": "BUG-2D1F509AB507",
  "title": "REINFORCE++ resets reward-to-go across multi-turn tool observations",
  "target": "verl",
  "validation_mode": "real_target_boundary_hook",
  "target_boundaries": [
    "verl.trainer.ppo.core_algos.compute_reinforce_plus_plus_outcome_advantage"
  ],
  "requires_model_download": false,
  "requires_dataset_download": false,
  "environment": {
    "VERL_REPO": "path to the veRL checkout under test",
    "OUTPUT_DIR": "directory where validation JSON should be written",
    "EXPECTATION": "confirmed for unpatched, not_triggered for fixed",
    "PYTHON": "python3"
  },
  "constructed_scenario": {
    "algorithm.adv_estimator": "reinforce_plus_plus",
    "gamma": 0.5,
    "tokens": [
      "model token before tool observation",
      "tool observation token with response_mask=0",
      "later model token with reward"
    ],
    "token_level_rewards": [[0.0, 0.0, 1.0]],
    "response_mask": [[1.0, 0.0, 1.0]]
  },
  "expected_unpatched": {
    "status": "confirmed",
    "returns": [[0.0, 0.5, 1.0]],
    "finding": "the internal response_mask=0 tool observation reset the running return before the earlier model token"
  },
  "expected_fixed": {
    "status": "not_triggered",
    "returns": [[0.5, 1.0, 1.0]],
    "finding": "the internal response_mask=0 tool observation was skipped without resetting reward-to-go"
  },
  "outputs": {
    "validation": "training_signal_validation.json"
  }
}
```

## Validation Runner

```bash
#!/usr/bin/env bash
set -euo pipefail

: "${VERL_REPO:?Set VERL_REPO to the veRL checkout under test}"
SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
OUTPUT_DIR="${OUTPUT_DIR:-${SCRIPT_DIR}/output}"
PYTHON="${PYTHON:-python3}"
EXPECTATION="${EXPECTATION:-any}"

mkdir -p "${OUTPUT_DIR}"

export VERL_REPO OUTPUT_DIR EXPECTATION
export PYTHONPATH="${VERL_REPO}${PYTHONPATH:+:${PYTHONPATH}}"

"${PYTHON}" "${SCRIPT_DIR}/rfpp_multiturn_returns_hook.py"
```

Example commands:

```bash
VERL_REPO=/path/to/unpatched/verl \
OUTPUT_DIR=/tmp/BUG-2D1F509AB507/unpatched \
EXPECTATION=confirmed \
artifacts/manual-repro/BUG-2D1F509AB507/run_validation.sh

VERL_REPO=/path/to/fixed/verl \
OUTPUT_DIR=/tmp/BUG-2D1F509AB507/fixed \
EXPECTATION=not_triggered \
artifacts/manual-repro/BUG-2D1F509AB507/run_validation.sh
```

## Boundary Hook

```python
from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path
from types import SimpleNamespace

import torch


BUG_ID = "BUG-2D1F509AB507"
TITLE = "REINFORCE++ resets reward-to-go across multi-turn tool observations"


def _git_commit(repo: Path) -> str | None:
    try:
        return subprocess.check_output(["git", "-C", str(repo), "rev-parse", "HEAD"], text=True).strip()
    except Exception:
        return None


def _tolist(tensor: torch.Tensor) -> list:
    return tensor.detach().cpu().tolist()


def _write_json(path: Path, payload: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")


def main() -> None:
    repo = Path(os.environ["VERL_REPO"]).resolve()
    output_dir = Path(os.environ["OUTPUT_DIR"]).resolve()
    expectation = os.environ.get("EXPECTATION", "any")

    if str(repo) not in sys.path:
        sys.path.insert(0, str(repo))

    from verl.trainer.ppo.core_algos import compute_reinforce_plus_plus_outcome_advantage

    token_level_rewards = torch.tensor([[0.0, 0.0, 1.0]], dtype=torch.float32)
    response_mask = torch.tensor([[1.0, 0.0, 1.0]], dtype=torch.float32)
    gamma = 0.5

    advantages, returns = compute_reinforce_plus_plus_outcome_advantage(
        token_level_rewards=token_level_rewards,
        response_mask=response_mask,
        config=SimpleNamespace(gamma=gamma),
    )

    observed_returns = _tolist(returns)
    unpatched_returns = [[0.0, 0.5, 1.0]]
    fixed_returns = [[0.5, 1.0, 1.0]]

    if torch.allclose(returns, torch.tensor(unpatched_returns, dtype=returns.dtype), atol=1e-6):
        status = "confirmed"
        finding = "the internal response_mask=0 tool observation reset running_return before the earlier model token"
    elif torch.allclose(returns, torch.tensor(fixed_returns, dtype=returns.dtype), atol=1e-6):
        status = "not_triggered"
        finding = "the internal response_mask=0 tool observation was skipped without resetting reward-to-go"
    else:
        status = "unexpected"
        finding = "RF++ returned an unrecognized reward-to-go pattern"

    payload = {
        "kind": "rl_sentinel_training_signal_validation",
        "bug_id": BUG_ID,
        "title": TITLE,
        "status": status,
        "finding": finding,
        "repo": str(repo),
        "commit": _git_commit(repo),
        "target_boundary": "verl.trainer.ppo.core_algos.compute_reinforce_plus_plus_outcome_advantage",
        "scenario": {
            "adv_estimator": "reinforce_plus_plus",
            "gamma": gamma,
            "token_level_rewards": [[0.0, 0.0, 1.0]],
            "response_mask": [[1.0, 0.0, 1.0]],
            "token_semantics": [
                "model token before tool observation",
                "tool observation token with response_mask=0",
                "later model token with final reward",
            ],
        },
        "observed": {
            "returns": observed_returns,
            "advantages": _tolist(advantages),
            "earlier_model_token_return": float(returns[0, 0].item()),
            "tool_observation_token_return": float(returns[0, 1].item()),
            "later_model_token_return": float(returns[0, 2].item()),
        },
        "expected": {
            "unpatched": {
                "status": "confirmed",
                "returns": unpatched_returns,
            },
            "fixed": {
                "status": "not_triggered",
                "returns": fixed_returns,
            },
        },
    }

    output_path = output_dir / "training_signal_validation.json"
    _write_json(output_path, payload)
    print(json.dumps(payload, indent=2))

    if expectation != "any" and status != expectation:
        raise SystemExit(f"expected status {expectation!r}, got {status!r}")


if __name__ == "__main__":
    main()
```

## Validation Output

Unpatched veRL:

```json
{
  "bug_id": "BUG-2D1F509AB507",
  "status": "confirmed",
  "commit": "5b8986bb94237163c603ed9be08d2f73c3e5d0ba",
  "observed": {
    "returns": [[0.0, 0.5, 1.0]],
    "earlier_model_token_return": 0.0,
    "tool_observation_token_return": 0.5,
    "later_model_token_return": 1.0
  }
}
```

Fixed branch:

```json
{
  "bug_id": "BUG-2D1F509AB507",
  "status": "not_triggered",
  "commit": "1420d8567888ab3c3056fd8d42eded886a5e35b5",
  "observed": {
    "returns": [[0.5, 1.0, 1.0]],
    "earlier_model_token_return": 0.5,
    "tool_observation_token_return": 1.0,
    "later_model_token_return": 1.0
  }
}
```

## Tests

```bash
python3 -m pytest -q \
  tests/trainer/ppo/test_core_algos_on_cpu.py::test_multi_turn_reinforce_plus_plus_returns_skip_observation_tokens \
  tests/trainer/ppo/test_core_algos_on_cpu.py::test_multi_turn_compute_gae_advantage_return
python3 -m pytest -q tests/trainer/ppo/test_core_algos_on_cpu.py
python3 -m ruff check verl/trainer/ppo/core_algos.py tests/trainer/ppo/test_core_algos_on_cpu.py
python3 -m ruff format --check verl/trainer/ppo/core_algos.py tests/trainer/ppo/test_core_algos_on_cpu.py
```

Results:

- `2 passed, 1 warning in 6.73s`
- `23 passed, 1 warning in 7.84s`
- `All checks passed!`
- `2 files already formatted`

## API Impact

No public API change. This only changes how RF++ carries reward-to-go through
internal non-trainable tool observation tokens.

AI assistance: this local draft and patch were prepared with Codex assistance
and reviewed in the local workspace.

### CODE REVIEW

**Review by wuxibin89 — DISMISSED** [review]

**Comment by haoyang9804** [comment]
@wuxibin89 hi Xibin, since the last commit triggered some ci failure, i fix the issue in my branch and re-commit it. Please take another look. Thx
